[Package] Docs and Pack / Unpack Artifact Types Separation (#9321)
mlrun/mlrun: docs/concepts/packagers/index.md

(packagers)=
# Packagers

Packagers are the recommended way to move data in and out of MLRun functions.
Instead of manually handling `DataItem`s, artifact uploads, and serialization code,
you write standard Python functions with type hints and returning values — and MLRun
automatically handles input parsing, output logging, and artifact creation.

```{toctree}
:maxdepth: 1

packagers-tutorial
custom_packagers/index
```

## What are packagers?

> Writing a function locally and running it remotely should feel identical. Locally, your

Reviewer: ```suggestion
Writing a function locally and running it remotely should feel identical. For example, running locally your
```

void. 

Packagers bridge this gap: they serialize inputs before they reach your code and
capture outputs after it runs, handling all the MLRun-specific I/O behind the scenes so
your function stays pure Python regardless of where it executes.

Every MLRun function has two I/O touch-points:

1. **Input parsing** — automatically cast `DataItem` inputs to the type-hinted Python type (e.g. `pd.DataFrame`, 
   `np.ndarray`, `dict`).
2. **Output logging** — automatically serialize, log, and upload returned objects as artifacts or results based on log 
   hints.

The flow looks like this:

**Input flow:** `inputs={"data": "store://..."}` → `DataItem` → packager `unpack()` → typed Python object → 
your function

**Output flow:** your function `return` → Python object → packager `pack()` → `Artifact` / `Result` → artifact store

## Why use packagers?

Packagers offer several advantages over manual artifact handling and the legacy context-based API.

### Better and faster learning curve

With packagers you don't need to learn about `Artifact`s, `DataItem`s, or the MLRun context object. You write standard 

> Python with type hints and returning values — MLRun integrates into your existing code without changing it.

Reviewer: ```suggestion
Python with type hints and returning values — MLRun wraps your existing code without changing it.
```

    # Log outputs manually
    context.log_result("row_count", row_count)
    context.log_dataset("cleaned_data", df=cleaned, format="parquet")
```

**After** — with packagers:

```python
import pandas as pd

def clean_data(raw_data: pd.DataFrame) -> tuple[int, pd.DataFrame]:
    cleaned = raw_data.dropna().drop_duplicates()
    return len(cleaned), cleaned
```

The function is pure Python — no MLRun imports, no manual serialization. When you run it with:

```python
fn.run(
    handler="clean_data",
    inputs={"raw_data": "store://my-raw-data"},
    returns=["row_count", "cleaned_data : dataset"],
)
```

MLRun automatically converts the `DataItem` to a DataFrame on input and logs the row count as a result and the
cleaned DataFrame as a dataset artifact on output.

### Uniformity of artifacts between users and projects

ML engineers can define how artifacts are serialized once and have that convention enforced
across every development notebook, CI pipeline, and production project in the organization.
Because packagers standardize the serialization format, artifacts become truly portable — a
DataFrame logged in one project can be consumed by a function in a completely different

> project without conversion steps or format mismatches.

Reviewer: ```suggestion
project without conversion steps or format mismatches (of course, given access is allowed across these projects).
```

In a pipeline, functions don't need to agree on file formats or know about MLRun's artifact
API. The producer just `return`s the object and the consumer receives it as a typed
parameter:

```python
# Producer — returns a DataFrame
def prepare_data(raw: pd.DataFrame) -> pd.DataFrame:
    return raw.dropna()

# Consumer — receives a DataFrame directly
def train_model(data: pd.DataFrame): ...
```

No manual `data_item.as_df()` calls, no format negotiation — the same artifact flows
cleanly between functions, projects, and teams.

### Automatic distribution across workers (coming soon)

Running at scale means dealing with multiple workers, jobs, nodes, resource allocation
(GPUs, memory), and a distribution framework — concerns that normally require significant
code changes. 

Packagers are the key to seamless distribution: because they know exactly how
to serialize and deserialize every argument, MLRun can automatically split inputs across
workers on different jobs and nodes, assign the right resources to each, and gather the
outputs back into a single result — all without touching your function code. A function

> with typed I/O can move from a single machine to a distributed cluster with zero changes.

Reviewer: Would remove the future-looking statements here - let's document what we have.

### Adaptive to user needs

MLRun provides common built-in packagers with rich options and configurations. For
example, you can control the output format with a single log hint:

>     returns=['data : dataset[format="parquet"]']

Reviewer: I would add an example of using the new `LogHint` as well as using the "legacy" format.

Author: I would rather leave the "legacy" dict format as it has no real "legacy" :)

instead of writing manual pandas I/O and artifact construction.

Beyond built-in packagers, MLRun supports **custom packagers** that you write and
register in your project to handle domain-specific types. Custom packagers imported

> from the MLRun Hub are coming soon.

Reviewer: Again, would not document things that are not there yet.

artifact type, labels, extra data, metrics, and more:

```python
from mlrun import LogHint

returns = [
    LogHint(key="model", artifact_type="model", labels={"version": "1"}),
    LogHint(key="data", artifact_type="dataset", packing_kwargs={"format": "csv"}),
]
```

A `LogHint` has the following fields:

| Field | Type | Description |
|-------|------|-------------|
| `key` | `str` | **Required.** The artifact key to log the object under. |
| `artifact_type` | `str \| None` | The artifact type (e.g. `"dataset"`, `"model"`, `"result"`). If `None`, the packager's default is used. |
| `tag` | `str` | Tag for the artifact. Default: `""`. |
| `itemized` | `bool \| int` | Unbundling control. `False` (default): log as one artifact. `True`: fully unbundle. `int`: unbundle to that depth. |
| `packing_kwargs` | `dict` | Extra keyword arguments passed to the packager's `pack_<type>()` method (e.g. `{"format": "parquet"}`). |
| `labels` | `dict[str, str]` | Labels to add to the logged artifact. |
| `extra_data` | `dict` | Extra data to attach to the artifact. Use `...` (Ellipsis) as a value to link to another package by key. |

> | `metrics` | `dict` | Metrics to log alongside a model artifact. Use `...` (Ellipsis) as a value to link to another package by key. |

Reviewer: It's hard to understand what the `...` description means. What do you mean by "value to link to another package"?

Author: lets say I:
`return my_model, some_results, loss_plot, test_dataset`

I would want `some_results` to be in the `my_model`'s metrics, and all the rest in `my_model`'s extra data. So I can in the log hint specify all the keys I want with an `...` (python common char for fill me or will be filled) in the metrics and extra data dicts.

I'll try to explain it better in the docs.

Author: see Linknig artifacts. at `packagers-overview.md`
